infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java

```diff
@@ -0,0 +1,211 @@
+  static final String ANNOTATION_ENV_NAME = ANNOTATION_PREFIX + "/" + "env-name";
+  static final String ANNOTATION_ENV_NAME_TEMPLATE = ANNOTATION_PREFIX + "/%s_" + "env-name";
+  static final String ANNOTATION_MOUNT_PATH = ANNOTATION_PREFIX + "/" + "mount-path";
+
+  private final Map<String, String> secretLabels;
+
+  @Inject
+  public SecretAsContainerResourceProvisioner(
+      @Named("che.workspace.provision.secret.labels") String[] labels) {
+    this.secretLabels =
+        Arrays.stream(labels)
+            .map(item -> item.split("=", 2))
+            .collect(toMap(p -> p[0], p -> p.length == 1 ? "" : p[1]));
+  }
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    LabelSelector selector = new LabelSelectorBuilder().withMatchLabels(secretLabels).build();
+    for (Secret secret : namespace.secrets().get(selector)) {
+      String targetContainerName =
+          secret.getMetadata().getAnnotations().get(ANNOTATION_TARGET_CONTAINER);
+      if ("env".equalsIgnoreCase(secret.getMetadata().getAnnotations().get(ANNOTATION_MOUNT_AS))) {
+        mountAsEnv(env, secret, targetContainerName);
+      } else {
+        mountAsFile(env, secret, targetContainerName);
+      }
+    }
+  }
+
+  private void mountAsEnv(E env, Secret secret, String targetContainerName)
+      throws InfrastructureException {
+    for (PodData podData : env.getPodsData().values()) {
+      if (!podData.getRole().equals(PodRole.DEPLOYMENT)) {
+        continue;
+      }
+      for (Container container : podData.getSpec().getContainers()) {
+        if (targetContainerName != null && !container.getName().equals(targetContainerName)) {
+          continue;
+        }
+        for (Entry<String, String> entry : secret.getData().entrySet()) {
+          final String mountEnvName = envName(secret, entry.getKey());
+          container
+              .getEnv()
+              .add(
+                  new EnvVarBuilder()
+                      .withName(mountEnvName)
+                      .withValueFrom(
+                          new EnvVarSourceBuilder()
+                              .withSecretKeyRef(
+                                  new SecretKeySelectorBuilder()
+                                      .withName(secret.getMetadata().getName())
+                                      .withKey(entry.getKey())
+                                      .build())
+                              .build())
+                      .build());
+        }
+      }
+    }
+  }
+
+  private void mountAsFile(E env, Secret secret, String targetContainerName)
+      throws InfrastructureException {
+
+    final String mountPath = secret.getMetadata().getAnnotations().get(ANNOTATION_MOUNT_PATH);
+    if (mountPath == null) {
+      throw new InfrastructureException(
+          format(
+              "Unable to mount secret '%s': it has to be mounted as file, but mountPath is not specified."
```

Comment:
Unable to mount secret '%s': It is configured to be mounted as a file but the mount path was not specified. Please define the '%s' annotation on the secret to specify it.